velox/common/memory/MemoryAllocator.cpp

```diff
@@ -271,7 +271,14 @@ bool MemoryAllocator::allocateContiguous(
     success = cache()->makeSpace(
         pagesToAcquire(numPages, numCollateralPages),
         [&](Allocation& acquired) {
+          const auto numPages = acquired.numPages();
           freeNonContiguous(acquired);
+          // We must advise pages away to get the OS to have more
+          // pages available for a contiguous allocation.
+          // freeNonContiguous only deallocates pages but leaves them used by
+          // the OS which can result in allocation failures even though the
+          // allocator technically has plenty of pages available for use.
+          unmap(numPages);
```

Comment:
Thank you for your comment. I did look at that function.
```
  const auto totalMaps =
      numMapped_.fetch_add(newMappedNeeded) + newMappedNeeded;
  if (totalMaps <= capacity_) {
    // We are not at capacity. No need to advise away.
    return true;
  }
```

This does return true in the following log where pages are not advised away in this code path
```
20240524 01:10:20.542948  3709 MmapAllocator.cpp:348] [MEM] Mmap failed with 2097152 pages use MmapArena false, errno Cannot allocate memory, Mmap Allocator: Memory Allocator[MMAP total capacity 238.00GB free capacity 14.75GB allocated pages 58522736 mapped pages 59502519 external mapped pages 7251976
[size 1: 16864(65MB) allocated 24061 mapped]
[size 2: 13166(102MB) allocated 233443 mapped]
[size 4: 16179(252MB) allocated 20477 mapped]
[size 8: 17990(562MB) allocated 27175 mapped]
[size 16: 7327(457MB) allocated 22944 mapped]
[size 32: 17154(2144MB) allocated 21992 mapped]
[size 64: 23750(5937MB) allocated 23823 mapped]
[size 128: 25972(12986MB) allocated 26014 mapped]
[size 256: 177767(177767MB) allocated 177871 mapped]
]
E20240524 01:10:20.553850  3709 MmapAllocator.cpp:348] [MEM] Mmap failed with 2097152 pages use MmapArena false, errno Cannot allocate memory, Mmap Allocator: Memory Allocator[MMAP total capacity 238.00GB free capacity 24.36GB allocated pages 56004605 mapped pages 59502519 external mapped pages 7251976
[size 1: 15903(62MB) allocated 24061 mapped]
[size 2: 12567(98MB) allocated 233443 mapped]
[size 4: 15268(238MB) allocated 20477 mapped]
[size 8: 16569(517MB) allocated 27175 mapped]
[size 16: 7057(441MB) allocated 22944 mapped]
[size 32: 15822(1977MB) allocated 21992 mapped]
[size 64: 22544(5636MB) allocated 23823 mapped]
[size 128: 24841(12420MB) allocated 26014 mapped]
[size 256: 169048(169048MB) allocated 177871 mapped]
]
E20240524 01:10:20.561141  3709 MmapAllocator.cpp:348] [MEM] Mmap failed with 2097152 pages use MmapArena false, errno Cannot allocate memory, Mmap Allocator: Memory Allocator[MMAP total capacity 238.00GB free capacity 32.64GB allocated pages 53834885 mapped pages 59502519 external mapped pages 7251976
[size 1: 15319(59MB) allocated 24061 mapped]
[size 2: 12185(95MB) allocated 233443 mapped]
[size 4: 14699(229MB) allocated 20477 mapped]
[size 8: 15807(493MB) allocated 27175 mapped]
[size 16: 6797(424MB) allocated 22944 mapped]
[size 32: 15036(1879MB) allocated 21992 mapped]
[size 64: 21672(5418MB) allocated 23823 mapped]
[size 128: 23593(11796MB) allocated 26014 mapped]
[size 256: 161567(161567MB) allocated 177871 mapped]
]
...
```

However, after looking at this again perhaps the calculation is incorrect when to decide to advise away.


It will only trigger if we run up to capacity and more pages are needed. The case we are trying to solve is where this is not triggered. The `totalMaps` doesn't exceed the capacity.

In this specific case we have 

`capacity_ == 62390272` (238GiB)
`numMapped == 59502519`
`newMappedNeed == 2097152`

In the sum `59502519 + 2097152 < 62390272` hence nothing is advised away which results in continuous allocation failures until it exceeds the retry limit and the query fails.

What is not considered in the check are the externally mapped pages for other contiguous allocations that necessitate the cache deallocations.

I suppose another option is to modify the check to
```
  const auto totalMaps =
      numMapped_.fetch_add(newMappedNeeded) + numExternalMapped_ + newMappedNeeded;
  if (totalMaps <= capacity_) {
    // We are not at capacity. No need to advise away.
    return true;
  }
```
in order to make sure we take the externally mapped pages into account when checking if free pages from previous noncontiguous allocations can be advised away.

Reply:
@majetideepak suggested the change to the check is better because it would help with scenarios without the async cache. I will make the change and add a test to check the advise away.

Reply:
@majetideepak @xiaoxmeng 

For some reason I though that `numMapped_` didn't include `numExternalMapped_` but that isn't the case. `numExternalMapped_` is incremented in `allocateContiguousImpl` and `numMapped_` is incremented in `ensureEnoughMappedPages` with the amount of contiguous requested pages.

That means my last comment was incorrect and my original point stands. We can have situations where we won't advise away even though we cannot allocate unless we do so.

Calculation: 
`52250543` pages in size classes  and `7251976` external mapped pages sum up to `59502519` pages total for `numMapped_`.